exporter/clickhouselogsexporter/exporter.go

```diff
@@ -117,10 +119,88 @@ func (e *clickhouseLogsExporter) Shutdown(_ context.Context) error {
 	return nil
 }
 
+func (e *clickhouseLogsExporter) getLogsTTLSeconds(ctx context.Context) (int, error) {
+	type DBResponseTTL struct {
+		EngineFull string `ch:"engine_full"`
+	}
+
+	var delTTL int = -1
+
+	var dbResp []DBResponseTTL
+	q := fmt.Sprintf("SELECT engine_full FROM system.tables WHERE name='%s' and database='%s'", tableName, databaseName)
+	err := e.db.Select(ctx, &dbResp, q)
+	if err != nil {
+		return delTTL, err
+	}
+	if len(dbResp) == 0 {
+		return delTTL, fmt.Errorf("ttl not found")
+	}
+
+	deleteTTLExp := regexp.MustCompile(`toIntervalSecond\(([0-9]*)\)`)
+
+	m := deleteTTLExp.FindStringSubmatch(dbResp[0].EngineFull)
+	if len(m) > 1 {
+		seconds_int, err := strconv.Atoi(m[1])
+		if err != nil {
+			return delTTL, nil
+		}
+		delTTL = seconds_int
+	}
+
+	return delTTL, nil
+}
+
+func (e *clickhouseLogsExporter) removeOldLogs(ctx context.Context, ld plog.Logs) error {
+	// get the TTL.
+	ttL, err := e.getLogsTTLSeconds(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if logs contains timestamp before acceptedDateTime, it will be rejected
+	acceptedDateTime := time.Now().Add(-(time.Duration(ttL) * time.Second))
+
+	removeLog := func(log plog.LogRecord) bool {
+		t := log.Timestamp().AsTime()
+		return t.Unix() < acceptedDateTime.Unix()
+	}
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		logs := ld.ResourceLogs().At(i)
+		for j := 0; j < logs.ScopeLogs().Len(); j++ {
+			logs.ScopeLogs().At(j).LogRecords().RemoveIf(removeLog)
```

Comment:
I forgot what is the default state. Is there any chance of panic in our setup?

Reply:
Didn't get you here

Reply:
`RemoveIf` mutates the own state. If it's state is immutable then it panics. You can check the implementation of `RemoveIf`. I just checked this shouldn't have any issue in our case.